lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -507,16 +598,204 @@ bool VerifySignatureMatches(llvm::Module *pModule, DXIL::SignatureKind SigKind,
 }
 
+struct SimplePSV {
+  uint32_t PSVRuntimeInfoSize = 0;
+  uint32_t PSVNumResources = 0;
+  uint32_t PSVResourceBindInfoSize = 0;
+  uint32_t StringTableSize = 0;
+  const char *StringTable = nullptr;
+  uint32_t SemanticIndexTableEntries = 0;
+  const uint32_t *SemanticIndexTable = nullptr;
+  uint32_t PSVSignatureElementSize = 0;
+  const PSVRuntimeInfo1 *RuntimeInfo1 = nullptr;
+  bool IsValid = true;
+  SimplePSV(const void *pPSVData, uint32_t PSVSize) {
+
+#define INCREMENT_POS(Size)                                                    \
+  Offset += Size;                                                              \
+  if (Offset > PSVSize) {                                                      \
+    IsValid = false;                                                           \
+    return;                                                                    \
+  }
+
+    uint32_t Offset = 0;
+    PSVRuntimeInfoSize = GetUint32AtOffset(pPSVData, 0);
+    INCREMENT_POS(4);
+    if (PSVRuntimeInfoSize >= sizeof(PSVRuntimeInfo1))
+      RuntimeInfo1 =
+          (const PSVRuntimeInfo1 *)(GetPtrAtOffset(pPSVData, Offset));
+    INCREMENT_POS(PSVRuntimeInfoSize);
+
+    PSVNumResources = GetUint32AtOffset(pPSVData, Offset);
+    INCREMENT_POS(4);
+    if (PSVNumResources > 0) {
+      PSVResourceBindInfoSize = GetUint32AtOffset(pPSVData, Offset);
+      // Increase the offset for the resource bind info size.
+      INCREMENT_POS(4);
+      // Increase the offset for the resource bind info.
+      INCREMENT_POS(PSVNumResources * PSVResourceBindInfoSize);
+    }
+    if (RuntimeInfo1) {
+      StringTableSize = GetUint32AtOffset(pPSVData, Offset);
+      INCREMENT_POS(4);
+      // Make sure StringTableSize is aligned to 4 bytes.
+      if ((StringTableSize & 3) != 0) {
+        IsValid = false;
+        return;
+      }
+      if (StringTableSize) {
+        StringTable = GetPtrAtOffset(pPSVData, Offset);
+        INCREMENT_POS(StringTableSize);
+      }
+      SemanticIndexTableEntries = GetUint32AtOffset(pPSVData, Offset);
+      INCREMENT_POS(4);
+      if (SemanticIndexTableEntries) {
+        SemanticIndexTable =
+            (const uint32_t *)(GetPtrAtOffset(pPSVData, Offset));
+        INCREMENT_POS(SemanticIndexTableEntries * 4);
+      }
+      if (RuntimeInfo1->SigInputElements || RuntimeInfo1->SigOutputElements ||
+          RuntimeInfo1->SigPatchConstOrPrimElements) {
+        PSVSignatureElementSize = GetUint32AtOffset(pPSVData, Offset);
+        INCREMENT_POS(4);
+        uint32_t PSVNumSignatures = RuntimeInfo1->SigInputElements +
+                                    RuntimeInfo1->SigOutputElements +
+                                    RuntimeInfo1->SigPatchConstOrPrimElements;
+        INCREMENT_POS(PSVNumSignatures * PSVSignatureElementSize);
+      }
+      if (RuntimeInfo1->UsesViewID) {
+        for (unsigned i = 0; i < DXIL::kNumOutputStreams; i++) {
+          uint32_t SigOutputVectors = RuntimeInfo1->SigOutputVectors[i];
```

Comment:
Do we also need to account for incrementing the position for SigInputVectors? Are they in the string table? I don't see any place where the position is incremented based on the presence of SigInputVectors like it's done here for SigOutputVectors.

Reply:
SigOutputVectors is an array of 4 for the case Geometry Shader might have 4 streams, each stream has its own output.
SigInputVectors is not an array, so not need to index for SigInputVectors.